testng-core/src/test/java/test/timeout/github2672/TimeoutStacktraceTest.java

```diff
@@ -0,0 +1,41 @@
+package test.timeout.github2672;
+
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+import java.util.Arrays;
+import org.testng.ITestListener;
+import org.testng.ITestResult;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+import org.testng.internal.thread.ThreadTimeoutException;
+
+@Listeners(TimeoutStacktraceTest.class)
+public class TimeoutStacktraceTest implements ITestListener {
+
+  @Test(
+      description = "testTimeoutStacktrace",
+      expectedExceptions = ThreadTimeoutException.class,
+      timeOut = 100)
+  public void testTimeoutStacktrace() throws InterruptedException {
+    Thread.sleep(1000);
+  }
+
+  private static void checkResults(ITestResult result) {
+    Throwable testError = result.getThrowable();
+    assertTrue(testError instanceof ThreadTimeoutException);
+    assertTrue(
```

Comment:
It is a bad practice to have exceptions or assertions in a listener.
Instead, you can store results and assert them into the test.